File: reference_pipeline/R/validate_flora.R
# FLoRA structural validation
#
# Checks data quality of output/flora.csv and produces a Markdown report
# with individual checkboxes per issue. Designed to be posted to a GitHub Issue;
# resolved issues disappear on the next run.
#
# False positive suppression: check an item in the GH issue, and the next
# workflow run will save it to output/flora_validation_suppressions.csv
# and exclude it from future reports.
#
# Usage: Rscript R/validate_flora.R

library(readr)
library(dplyr)
library(stringr)
library(glue)
library(here)

source(here("R", "outcome_vocabulary.R"))

is_doi <- function(x) grepl("^10\\.[0-9]{4,}/\\S+$", x)

SUPPRESSIONS_FILE <- here("output", "flora_validation_suppressions.csv")

# ---------------------------------------------------------------------------
# Parse a GH issue body to extract checked items as suppressions.
# Returns a tibble with check_name and item_id columns.
# ---------------------------------------------------------------------------
parse_issue_suppressions <- function(issue_body_file) {
  if (!file.exists(issue_body_file)) {
    return(tibble(check_name = character(), item_id = character()))
  }
  lines <- readLines(issue_body_file, warn = FALSE)
  current_section <- ""
  results <- list()

  for (line in lines) {
    # Stop parsing at the "Checks passed" section (those are always [x])
    if (str_detect(line, "^\\*\\*Checks passed")) break

    # Bold section header: **Check name** (N issues):
    if (str_detect(line, "^\\*\\*[^*]+\\*\\*")) {
      current_section <- str_extract(line, "(?<=^\\*\\*)[^*]+(?=\\*\\*)")
    }
    # <details><summary><b>Check name</b> ...
    if (str_detect(line, "<b>[^<]+</b>")) {
      current_section <- str_extract(line, "(?<=<b>)[^<]+(?=</b>)")
    }
    # Checked item: - [x] ...
    if (str_detect(line, "^- \\[x\\]") && nzchar(current_section)) {
      item_text <- str_replace(line, "^- \\[x\\] ", "")
      item_id <- extract_item_id(item_text)
      results <- c(results, list(tibble(
        check_name = current_section, item_id = item_id
      )))
    }
  }

  if (length(results) > 0) bind_rows(results) else {
    tibble(check_name = character(), item_id = character())
  }
}

# Extract the stable ID part from an item label.
# Structural: "row_id: details" -> "row_id", or just "row_id"
# Network DOI/URL: "`value`" -> "value"
# Network retracted: "**doi_r**: `value` -- id" -> "value"
extract_item_id <- function(item_text) {
  sapply(item_text, function(txt) {
    if (str_detect(txt, "^`")) {
      # Network: `doi_or_url`
      str_extract(txt, "(?<=^`)[^`]+")
    } else if (str_detect(txt, "^\\*\\*")) {
      # Network retracted: **doi_r**: `10.xxx` ...
      str_extract(txt, "(?<=`)[^`]+(?=`)")
    } else {
      # Structural: "id: details" or just "id"
      pos <- regexpr(": ", txt, fixed = TRUE)
      if (pos > 0) substr(txt, 1, pos - 1) else txt
    }
  }, USE.NAMES = FALSE)
}

# ---------------------------------------------------------------------------
# Main validation function
# ---------------------------------------------------------------------------
validate_flora_data <- function(data, suppressions = NULL) {

  # --- Row identifier (FLoRA has no id column) ---
  data <- data %>%
    mutate(id = paste(doi_o, coalesce(doi_r, url_r), sep = " | "))

  # --- Valid value sets ---
  # VALID_REPLICATION_OUTCOMES / VALID_REPRODUCTION_OUTCOMES come from
  # R/outcome_vocabulary.R, which both pipelines normalise against.
  VALID_TYPES <- c("replication", "reproduction")
  VALID_SOURCES <- c("COS", "replications", "reproductions")
  URL_COLS <- c("url_r", "oa_url_o", "oa_url_r")
  DETAILS_THRESHOLD <- 10
  current_year <- as.integer(format(Sys.Date(), "%Y"))

  # --- Storage ---
  all_checks <- list()
  n_suppressed <- 0L

  # --- Helpers ---
  truncate_for_label <- function(x, width = 80) {
    str_trunc(as.character(x), width = width, ellipsis = "...")
  }

  format_labels <- function(issue_df, detail_template = NULL, detail_fn = NULL) {
    if (nrow(issue_df) == 0) return(character())
    issue_df <- distinct(issue_df) %>% mutate(id = as.character(id))
    if (!is.null(detail_fn)) return(detail_fn(issue_df))
    if (!is.null(detail_template)) {
      details <- truncate_for_label(glue_data(issue_df, detail_template))
      return(unique(glue("{issue_df$id}: {details}")))
    }
    unique(issue_df$id)
  }

  add_check <- function(issue_df, check_name, fail_heading = NULL,
                        description = NULL,
                        detail_template = NULL, detail_fn = NULL) {
    issue_df <- distinct(issue_df) %>% mutate(id = as.character(id))
    if (nrow(issue_df) > 0) {
      items <- format_labels(issue_df, detail_template, detail_fn)
      # Filter out suppressed items (match by check_name or fail_heading)
      fh <- fail_heading %||% check_name
      if (!is.null(suppressions) && nrow(suppressions) > 0) {
        suppressed_ids <- suppressions$item_id[
          suppressions$check_name %in% c(check_name, fh)
        ]
        if (length(suppressed_ids) > 0) {
          keep <- !extract_item_id(items) %in% suppressed_ids
          n_suppressed <<- n_suppressed + sum(!keep)
          items <- items[keep]
        }
      }
      if (length(items) > 0) {
        all_checks[[check_name]] <<- list(
          status = "fail", items = items,
          fail_heading = fh, description = description
        )
      } else {
        all_checks[[check_name]] <<- list(status = "pass")
      }
    } else {
      all_checks[[check_name]] <<- list(status = "pass")
    }
  }

  # =========================================================================
  # Check 1: No 'NA'/'N/A' literal strings
  # =========================================================================
  issues <- data %>%
    select(id, where(is.character)) %>%
    tidyr::pivot_longer(-id, names_to = "column", values_to = "value",
                        values_drop_na = TRUE) %>%
    filter(value %in% c("NA", "N/A"))
  add_check(issues, "No 'NA'/'N/A' strings",
            fail_heading = "Literal 'NA'/'N/A' strings found",
            description = "Text columns contain the literal string 'NA' or 'N/A' instead of a proper missing value.",
            detail_template = "{column}='{value}'")

  # =========================================================================
  # Check 2: All links are valid URLs (start with http)
  # =========================================================================
  url_cols_present <- intersect(URL_COLS, names(data))
  issues <- data %>%
    select(id, all_of(url_cols_present)) %>%
    tidyr::pivot_longer(-id, names_to = "column", values_to = "value",
                        values_drop_na = TRUE) %>%
    filter(!str_starts(value, "http"))
  add_check(issues, "All links are valid URLs",
            fail_heading = "Invalid URLs (not starting with http)",
            description = "Values in url_r, oa_url_o, or oa_url_r that don't start with 'http'.",
            detail_template = "{column}='{value}'")

  # =========================================================================
  # Check 6: DOI mapped to conflicting references
  # =========================================================================
  clean_ref_text <- function(x) {
    x %>%
      str_to_lower() %>%
      str_replace_all("\\bdoi:\\s*10[./]\\S+", "") %>%
      str_replace_all("https?://(dx\\.)?doi\\.org/10[./]\\S+", "") %>%
      str_replace_all("study\\s*\\d+[a-zA-Z]?\\b", "") %>%
      str_squish()
  }

  min_norm_distance <- function(refs) {
    refs <- unique(refs[!is.na(refs)])
    if (length(refs) <= 1) return(0)
    combos <- combn(seq_along(refs), 2)
    dists <- apply(combos, 2, function(idx) {
      r1 <- refs[idx[1]]; r2 <- refs[idx[2]]
      as.numeric(adist(r1, r2)) / max(nchar(r1), nchar(r2), 1)
    })
    min(dists)
  }

  check_doi_ref_conflicts <- function(data, doi_col, ref_col, label) {
    df <- data %>%
      select(id, doi = !!sym(doi_col), ref = !!sym(ref_col)) %>%
      filter(!is.na(doi), !is.na(ref)) %>%
      mutate(ref_clean = clean_ref_text(ref))
    if (nrow(df) == 0) return(tibble())
    flagged_dois <- df %>%
      group_by(doi) %>%
      filter(n() > 1) %>%
      summarise(min_dist = min_norm_distance(ref_clean), .groups = "drop") %>%
      filter(min_dist > 0.2) %>%
      pull(doi)
    df %>% filter(doi %in% flagged_dois) %>% mutate(source = label)
  }

  issues <- bind_rows(
    check_doi_ref_conflicts(data, "doi_o", "apa_ref_o", "doi_o/apa_ref_o"),
    check_doi_ref_conflicts(data, "doi_r", "apa_ref_r", "doi_r/apa_ref_r")
  )
  add_check(issues, "DOI mapped to conflicting references",
    description = "The same DOI appears with substantially different APA reference text across rows (>20% edit distance). May indicate a wrong DOI or inconsistent reference formatting.",
    detail_fn = function(df) {
      df %>%
        mutate(id = as.character(id),
               ref_short = truncate_for_label(ref, 60),
               detail = glue("{doi} ({source}); ref='{ref_short}'"),
               detail = truncate_for_label(detail)) %>%
        {unique(glue("{.$id}: {.$detail}"))}
    }
  )

  # =========================================================================
  # Check 7: Outcome values valid
  # =========================================================================
  issues <- data %>%
    filter(
      (type == "replication" & !outcome %in% VALID_REPLICATION_OUTCOMES) |
      (type == "reproduction" & !outcome %in% VALID_REPRODUCTION_OUTCOMES) |
      is.na(outcome)
    )
  add_check(issues, "Outcome values valid",
            description = paste0(
              "Outcome values that don't match the allowed set for the row's type ",
              "(replication or reproduction). A value containing '", trimws(OUTCOME_CLASH_SEP),
              "' is a genuine outcome clash between rows merged during preprint ",
              "deduplication and needs resolving in the source coding sheet."),
            detail_template = "type={type}; outcome='{outcome}'")

  # =========================================================================
  # Check 8: Type values valid
  # =========================================================================
  issues <- data %>% filter(!type %in% VALID_TYPES)
  add_check(issues, "Type values valid",
            description = "Type must be 'replication' or 'reproduction'.",
            detail_template = "type='{type}'")

  # =========================================================================
  # Check 9: Source values valid
  # =========================================================================
  issues <- data %>% filter(!source %in% VALID_SOURCES)
  add_check(issues, "Source values valid",
            description = "Source must be 'COS', 'replications', or 'reproductions'.",
            detail_template = "source='{source}'")

  # =========================================================================
  # Check 10: Year ranges reasonable
  # =========================================================================
  issues <- data %>%
    filter(
      (!is.na(year_o) & (year_o < 1890 | year_o > current_year + 1)) |
      (!is.na(year_r) & (year_r < 1890 | year_r > current_year + 1))
    )
  add_check(issues, "Year ranges reasonable",
            description = "year_o or year_r is outside 1890\u2013current year + 1.",
            detail_template = "year_o={year_o}; year_r={year_r}")

  # =========================================================================
  # Check 11: year_r >= year_o (warning)
  # =========================================================================
  issues <- data %>%
    filter(!is.na(year_o), !is.na(year_r), year_r < year_o)
  add_check(issues, "year_r >= year_o",
            fail_heading = "Replication year before original year",
            description = "year_r < year_o. Could indicate swapped entries or a data-entry error.",
            detail_template = "year_o={year_o}; year_r={year_r}")

  # =========================================================================
  # Check 12: No exact duplicates
  # =========================================================================
  issues <- data %>%
    group_by(doi_o, doi_r) %>%
    filter(n() > 1) %>%
    ungroup()
  add_check(issues, "No exact duplicates",
            fail_heading = "Exact duplicates (by doi_o + doi_r)",
            description = "Rows that share the same doi_o and doi_r combination.")

  # =========================================================================
  # Check 13: Required fields present
  # =========================================================================
  issues <- data %>%
    filter(is.na(title_o) | is.na(title_r) | is.na(doi_o) |
           (is.na(doi_r) & is.na(url_r)))
  add_check(issues, "Required fields present",
            fail_heading = "Missing required fields",
            description = "title_o, title_r, and doi_o must be non-NA; at least one of doi_r or url_r must be present.")

  # =========================================================================
  # Check 14: DOI format valid
  # =========================================================================
  issues <- data %>%
    filter(
      (!is.na(doi_o) & !is_doi(doi_o)) |
      (!is.na(doi_r) & !is_doi(doi_r))
    )
  add_check(issues, "DOI format valid",
            fail_heading = "Invalid DOI format",
            description = "DOI does not match the expected pattern `10.NNNN/...`.",
            detail_template = "doi_o={doi_o}; doi_r={doi_r}")

  # =========================================================================
  # Build Markdown report (individual checkboxes per issue)
  # =========================================================================
  report_parts <- c(
    "## FLoRA Data Validation Report\n",
    glue("_Generated: {format(Sys.time(), '%Y-%m-%d %H:%M:%S')} | Dataset: {nrow(data)} rows_"),
    "",
    "Each row is identified as `doi_o | doi_r (or url_r)`, followed by check-specific details after `:`.",
    "Check an item to mark it as a false positive; it will be suppressed on the next run.\n"
  )

  fail_checks <- all_checks[sapply(all_checks, function(x) x$status == "fail")]
  pass_checks <- all_checks[sapply(all_checks, function(x) x$status == "pass")]

  if (length(fail_checks) > 0) {
    for (check_name in names(fail_checks)) {
      heading <- fail_checks[[check_name]]$fail_heading
      desc <- fail_checks[[check_name]]$description
      items <- fail_checks[[check_name]]$items
      n <- length(items)
      checkbox_lines <- paste0("- [ ] ", items)
      desc_line <- if (!is.null(desc)) glue("_{desc}_\n") else ""
      if (n <= DETAILS_THRESHOLD) {
        report_parts <- c(report_parts,
          glue("**{heading}** ({n} issue{ifelse(n > 1, 's', '')}):\n"),
          desc_line,
          checkbox_lines, "")
      } else {
        report_parts <- c(report_parts,
          glue("<details><summary><b>{heading}</b> ({n} issues)</summary>"),
          "",  # blank line required for GFM to render markdown inside HTML
          desc_line,
          checkbox_lines, "",
          "</details>", "")
      }
    }
  }

  if (length(pass_checks) > 0) {
    report_parts <- c(report_parts, "---\n", "**Checks passed:**")
    for (check_name in names(pass_checks)) {
      report_parts <- c(report_parts, glue("- [x] {check_name}"))
    }
    report_parts <- c(report_parts, "")
  }

  if (n_suppressed > 0) {
    report_parts <- c(report_parts,
      glue("_({n_suppressed} suppressed false positive{ifelse(n_suppressed > 1, 's', '')} not shown)_"))
  }

  has_issues <- length(fail_checks) > 0
  final_report <- paste(report_parts, collapse = "\n")
  list(report = final_report, has_issues = has_issues)
}

# === CLI entry point (only when run directly, not when sourced) ===
is_main_script <- function(script_name) {
  args <- commandArgs(trailingOnly = FALSE)
  script_arg <- args[grep("--file=", args)]
  if (length(script_arg) == 0) return(FALSE)
  script_path <- sub("--file=", "", script_arg)
  grepl(script_name, normalizePath(script_path, mustWork = FALSE), fixed = TRUE)
}

if (is_main_script("validate_flora.R") ||
    identical(Sys.getenv("FLORA_VALIDATE_RUN"), "1")) {
  input_file <- here("output", "flora.csv")
  if (!file.exists(input_file)) stop("flora.csv not found at: ", input_file)

  flora <- read_csv(input_file, show_col_types = FALSE)
  message("Read ", nrow(flora), " rows from flora.csv")

  # Load existing suppressions
  suppressions <- if (file.exists(SUPPRESSIONS_FILE)) {
    read_csv(SUPPRESSIONS_FILE, show_col_types = FALSE)
  } else {
    tibble(check_name = character(), item_id = character(),
           suppressed_date = character())
  }

  # Parse new suppressions from GH issue body (CI sets this env var)
  issue_body_file <- Sys.getenv("FLORA_ISSUE_BODY_FILE", "")
  if (nzchar(issue_body_file)) {
    new_supp <- parse_issue_suppressions(issue_body_file)
    if (nrow(new_supp) > 0) {
      new_supp$suppressed_date <- format(Sys.Date(), "%Y-%m-%d")
      suppressions <- bind_rows(suppressions, new_supp) %>%
        distinct(check_name, item_id, .keep_all = TRUE)
      write_excel_csv(suppressions, SUPPRESSIONS_FILE)
      message("Updated suppressions: ", nrow(new_supp), " new entries")
    }
  }

  if (nrow(suppressions) > 0) {
    message("Loaded ", nrow(suppressions), " suppressions")
  }

  result <- validate_flora_data(flora, suppressions = suppressions)
  cat(result$report, "\n")
}

File: validate_flora.py
"""validate_flora.py — structural validation of the FLoRA dataset.

A Python port of R/validate_flora.R. Runs the same checks against the dataset this
project now builds itself, and emits the same Markdown report: one checkbox per
issue, so a resolved issue simply disappears on the next run and a false positive
can be ticked to suppress it.

WHAT CHANGED IN THE PORT
------------------------
`source` values. Both the supplied notebook labels and the website's registry keys
are accepted, and preserved reference rows
must not become validation failures merely because their labels predate the site.

Suppressions live in the same CSV the R version uses
(`output/flora_validation_suppressions.csv`), in the same shape, so a suppression
ticked under either implementation is honoured by both.

Usage:
    python validate_flora.py                     # report to stdout
    python validate_flora.py --output report.md
    python validate_flora.py --fail-on-issues    # exit 1 if anything is flagged

Required environment variables:
    DATABASE_URL — PostgreSQL connection string
"""
import argparse
import csv
import os
import re
import sys
from datetime import datetime, timezone
from pathlib import Path

import pandas as pd
import psycopg2
import psycopg2.extras
from dotenv import load_dotenv

from console_encoding import use_utf8_output
from pipeline_logging import start as start_logging
from extractor_vocab import REPLICATION_OUTCOMES

load_dotenv()
use_utf8_output()

ROOT = Path(__file__).parent
SUPPRESSIONS_FILE = ROOT / "output" / "flora_validation_suppressions.csv"

VALID_SOURCES = {
    "entry_sheet_replications", "entry_sheet_reproductions",
    "fred_replication_success", "score_2025", "validated",
    "COS", "SCORE", "replications", "reproductions",
    "openalex", "openalex_snapshot", "i4r",
}
VALID_TYPES = {"replication", "reproduction"}
URL_COLUMNS = ["url_r", "oa_url_o", "oa_url_r", "url_o"]

# Above this many issues the section is folded into a <details> block, so one noisy
# check cannot bury every other one.
DETAILS_THRESHOLD = 10

DOI_RE = re.compile(r"^10\.[0-9]{4,}/\S+$")
NA_STRINGS = {"NA", "N/A", "na", "n/a", "NULL", "None"}


def _blank(value) -> bool:
    return (value is None or (isinstance(value, float) and pd.isna(value))
            or not str(value).strip())


def _text(value) -> str:
    return "" if _blank(value) else str(value).strip()


def row_id(row) -> str:
    """`doi_o | doi_r-or-url_r`, the identifier the R report uses.

    Kept even though our rows now carry a real flora_id, because a suppression
    ticked against the R report has to keep matching. flora_id is shown beside it.
    """
    right = _text(row.get("doi_r")) or _text(row.get("url_r"))
    return f"{_text(row.get('doi_o'))} | {right}"


def _edit_distance(a: str, b: str) -> int:
    """Levenshtein, as R's adist computes it."""
    if a == b:
        return 0
    if not a or not b:
        return max(len(a), len(b))
    previous = list(range(len(b) + 1))
    for i, ca in enumerate(a, 1):
        current = [i]
        for j, cb in enumerate(b, 1):
            current.append(min(previous[j] + 1, current[j - 1] + 1,
                               previous[j - 1] + (ca != cb)))
        previous = current
    return previous[-1]


def _normalised_distance(a: str, b: str) -> float:
    return _edit_distance(a, b) / max(len(a), len(b), 1)


_REF_DOI_RE = re.compile(r"\bdoi:\s*10[./]\S+|https?://(dx\.)?doi\.org/10[./]\S+", re.I)
_REF_STUDY_RE = re.compile(r"study\s*\d+[a-z]?\b", re.I)


def _clean_ref(value: str) -> str:
    text = _text(value).lower()
    text = _REF_DOI_RE.sub("", text)
    text = _REF_STUDY_RE.sub("", text)
    return " ".join(text.split())


# ── suppressions ──────────────────────────────────────────────────────────────

def load_suppressions(path: Path = SUPPRESSIONS_FILE) -> set:
    """(check_name, item_id) pairs previously ticked as false positives."""
    if not path.exists():
        return set()
    out = set()
    with path.open(encoding="utf-8", newline="") as fh:
        for row in csv.DictReader(fh):
            if row.get("check_name") and row.get("item_id"):
                out.add((row["check_name"].strip(), row["item_id"].strip()))
    return out


def item_id(label: str) -> str:
    """The stable part of an item label — everything before the first ': '."""
    text = label.strip()
    if text.startswith("`"):
        end = text.find("`", 1)
        return text[1:end] if end > 0 else text
    marker = text.find(": ")
    return text[:marker] if marker > 0 else text


# ── the checks ────────────────────────────────────────────────────────────────

def _check_na_strings(df):
    issues = []
    for _, row in df.iterrows():
        for column, value in row.items():
            if isinstance(value, str) and value.strip() in NA_STRINGS:
                issues.append(f"{row_id(row)}: {column}='{value.strip()}'")
    return issues


def _check_urls(df):
    issues = []
    for _, row in df.iterrows():
        for column in URL_COLUMNS:
            value = _text(row.get(column))
            if value and not value.lower().startswith("http"):
                issues.append(f"{row_id(row)}: {column}='{value}'")
    return issues


def _check_outcomes(df):
    from extractor_vocab import (normalize_axis_value, OUTCOME_COMPUTATION_VALUES,
                                 OUTCOME_ROBUSTNESS_VALUES)
    issues = []
    for _, row in df.iterrows():
        outcome = _text(row.get("outcome"))
        if not outcome:
            issues.append(f"{row_id(row)}: outcome is blank")
        elif row.get("type") == "replication" and outcome not in REPLICATION_OUTCOMES:
            issues.append(f"{row_id(row)}: type=replication; outcome='{outcome}'")
        elif row.get("type") == "reproduction":
            parts = [part.strip() for part in outcome.split(",")]
            try:
                valid = (len(parts) == 2
                         and normalize_axis_value("outcome_computation", parts[0]) in OUTCOME_COMPUTATION_VALUES
                         and normalize_axis_value("outcome_robustness", parts[1]) in OUTCOME_ROBUSTNESS_VALUES)
            except ValueError:
                valid = False
            if not valid:
                issues.append(f"{row_id(row)}: type=reproduction; outcome='{outcome}'")
    return issues


def _check_types(df):
    return [f"{row_id(r)}: type='{_text(r.get('type'))}'"
            for _, r in df.iterrows() if r.get("type") not in VALID_TYPES]


def _check_sources(df):
    return [f"{row_id(r)}: source='{_text(r.get('source'))}'"
            for _, r in df.iterrows() if r.get("source") not in VALID_SOURCES]


def _year(value):
    try:
        return int(float(str(value)))
    except (TypeError, ValueError):
        return None


def _check_year_ranges(df):
    current = datetime.now(timezone.utc).year
    issues = []
    for _, row in df.iterrows():
        for column in ("year_o", "year_r"):
            year = _year(row.get(column))
            if year is not None and (year < 1890 or year > current + 1):
                issues.append(f"{row_id(row)}: {column}={year}")
    return issues


def _check_year_order(df):
    issues = []
    for _, row in df.iterrows():
        year_o, year_r = _year(row.get("year_o")), _year(row.get("year_r"))
        if year_o and year_r and year_r < year_o:
            issues.append(f"{row_id(row)}: year_o={year_o}; year_r={year_r}")
    return issues


def _check_exact_duplicates(df):
    pairs = df[["doi_o", "doi_r"]].fillna("")
    duplicated = pairs.duplicated(keep=False) & (pairs["doi_o"] != "")
    return sorted({row_id(r) for _, r in df[duplicated].iterrows()})


def _check_required_fields(df):
    issues = []
    for _, row in df.iterrows():
        missing = [c for c in ("title_o", "title_r", "doi_o") if _blank(row.get(c))]
        if _blank(row.get("doi_r")) and _blank(row.get("url_r")):
            missing.append("doi_r or url_r")
        if missing:
            issues.append(f"{row_id(row)}: missing {', '.join(missing)}")
    return issues


def _check_doi_format(df):
    issues = []
    for _, row in df.iterrows():
        for column in ("doi_o", "doi_r"):
            value = _text(row.get(column))
            if value and not DOI_RE.match(value):
                issues.append(f"{row_id(row)}: {column}='{value}'")
    return issues


def _check_reference_conflicts(df):
    """The same DOI appearing with substantially different reference text."""
    issues = []
    for doi_column, ref_column in (("doi_o", "apa_ref_o"), ("doi_r", "apa_ref_r")):
        if ref_column not in df.columns:
            continue
        frame = df[[doi_column, ref_column]].dropna()
        for doi, group in frame.groupby(doi_column):
            refs = {_clean_ref(r) for r in group[ref_column] if _clean_ref(r)}
            if len(refs) < 2:
                continue
            refs = sorted(refs)
            worst = max(_normalised_distance(a, b)
                        for i, a in enumerate(refs) for b in refs[i + 1:])
            if worst > 0.2:
                issues.append(f"{doi}: {len(refs)} different reference strings "
                              f"({doi_column}/{ref_column})")
    return issues


CHECKS = [
    ("No 'NA'/'N/A' strings", "Literal 'NA'/'N/A' strings found", _check_na_strings,
     "A text column holds the literal string 'NA' instead of a missing value."),
    ("All links are valid URLs", "Invalid URLs (not starting with http)", _check_urls,
     "A url column holds something that is not a link."),
    ("Outcome values valid", "Invalid outcome values", _check_outcomes,
     "An outcome outside the allowed set for the row's type, or missing."),
    ("Type values valid", "Invalid type values", _check_types,
     "type must be 'replication' or 'reproduction'."),
    ("Source values valid", "Invalid source values", _check_sources,
     "source must identify a supported dataset input."),
    ("Year ranges reasonable", "Implausible years", _check_year_ranges,
     "year_o or year_r is outside 1890 to next year."),
    ("year_r >= year_o", "Replication year before original year", _check_year_order,
     "Could indicate swapped entries or a data-entry error."),
    ("No exact duplicates", "Exact duplicates (by doi_o + doi_r)",
     _check_exact_duplicates,
     "Rows sharing the same doi_o and doi_r after deduplication."),
    ("Required fields present", "Missing required fields", _check_required_fields,
     "title_o, title_r and doi_o must be present, plus doi_r or url_r."),
    ("DOI format valid", "Invalid DOI format", _check_doi_format,
     "DOI does not match 10.NNNN/..."),
    ("DOI mapped to conflicting references", "DOI with conflicting references",
     _check_reference_conflicts,
     "The same DOI appears with reference strings more than 20% apart."),
]


def validate(df, suppressions=None) -> dict:
    suppressions = suppressions or set()
    results, suppressed = [], 0

    for name, heading, fn, description in CHECKS:
        items = fn(df)
        kept = []
        for label in items:
            if (name, item_id(label)) in suppressions or (heading, item_id(label)) in suppressions:
                suppressed += 1
            else:
                kept.append(label)
        results.append({"name": name, "heading": heading, "description": description,
                        "items": kept, "passed": not kept})

    return {"results": results, "suppressed": suppressed,
            "has_issues": any(not r["passed"] for r in results), "rows": len(df)}


def render(report: dict) -> str:
    stamp = datetime.now(timezone.utc).strftime("%Y-%m-%d %H:%M:%S")
    out = [
        "## FLoRA Data Validation Report",
        "",
        f"_Generated: {stamp} UTC | Dataset: {report['rows']} rows_",
        "",
        "Each row is identified as `doi_o | doi_r (or url_r)`, followed by "
        "check-specific detail after `:`.",
        "Tick an item to mark it a false positive; it is suppressed on the next run.",
        "",
    ]

    failed = [r for r in report["results"] if not r["passed"]]
    passed = [r for r in report["results"] if r["passed"]]

    for result in failed:
        items = result["items"]
        n = len(items)
        lines = [f"- [ ] {label}" for label in items]
        if n <= DETAILS_THRESHOLD:
            out += [f"**{result['heading']}** ({n} issue{'s' if n != 1 else ''}):", "",
                    f"_{result['description']}_", ""] + lines + [""]
        else:
            out += [f"<details><summary><b>{result['heading']}</b> ({n} issues)</summary>",
                    "", f"_{result['description']}_", ""] + lines + ["", "</details>", ""]

    if passed:
        out += ["---", "", "**Checks passed:**"]
        out += [f"- [x] {r['name']}" for r in passed]
        out.append("")

    if report["suppressed"]:
        out.append(f"_({report['suppressed']} suppressed false positive"
                   f"{'s' if report['suppressed'] != 1 else ''} not shown)_")
    return "\n".join(out)


def load_dataset(cur):
    """The product as it is published, with the output contract's column names.

    Untitled rows are dropped here for the same reason the export drops them: the
    R notebook validates output/flora.csv, which has already been through Step 10.
    Reporting a row that was never published would put an item in the issue that
    nobody can resolve — it is already recorded in output/flora_export_log.csv and
    counted on the FLoRA tab.
    """
    import flora_registry
    import transform_sources
    frame = transform_sources.build(cur, verbose=False)
    frame = flora_registry.attach_ids(cur, frame)
    return transform_sources.drop_untitled(transform_sources.to_output_shape(frame))


def main() -> int:
    start_logging("validate")
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--output", type=Path,
                        help="write the report here instead of stdout")
    parser.add_argument("--input", type=Path,
                        help="validate this exact CSV instead of rebuilding from the database")
    parser.add_argument("--fail-on-issues", action="store_true",
                        help="exit 1 when anything is flagged (for CI)")
    args = parser.parse_args()

    if args.input:
        df = pd.read_csv(args.input, dtype=str, keep_default_na=False,
                         na_values=["NA", ""], encoding="utf-8-sig")
    else:
        database_url = os.environ.get("DATABASE_URL")
        if not database_url:
            print("DATABASE_URL is not set", file=sys.stderr)
            return 2
        conn = psycopg2.connect(database_url)
        try:
            cur = conn.cursor(cursor_factory=psycopg2.extras.RealDictCursor)
            df = load_dataset(cur)
        finally:
            conn.close()

    report = validate(df, load_suppressions())
    text = render(report)

    if args.output:
        args.output.parent.mkdir(parents=True, exist_ok=True)
        args.output.write_text(text, encoding="utf-8")
        print(f"report written to {args.output}")
    else:
        print(text)

    failed = sum(1 for r in report["results"] if not r["passed"])
    total = sum(len(r["items"]) for r in report["results"])
    print(f"\n{failed}/{len(report['results'])} check(s) flagged {total} issue(s)",
          file=sys.stderr)
    return 1 if (args.fail_on_issues and report["has_issues"]) else 0


if __name__ == "__main__":
    raise SystemExit(main())
